### PR TITLE
BUG: Fix bug related to how Neuroscan .cnt n_samples is read

### PR DESCRIPTION
**Pull request by withmywoessner**
#### Reference issue
#6537
#11802

#### What does this implement/fix?
Change how number of samples is read and verifies that they are correct.

#### Additional information
I need to upload new testing data that demonstrates this bug.

### CODE REVIEW

**Comment by withmywoessner** [comment]
@larsoner Should I update the testing data to include a file that was not read properly before this change? 

The documentation said that the header I am using was not reliable, but I think that may be because they read the n_samples as normal int when it should be an unsigned int.

The eeglab read_cnt function reads the n_samples properly and they also read it as an unsigned int.

**Comment by larsoner** [comment]
> Can I update the testing data to include a file that was not read properly before this change or even replace the one I already added.

Yes feel free! The file is small, right? The smaller the testing dataset the better, so might as well replace the old one assuming we can still test all the same stuff.

File: mne/io/cnt/cnt.py [hunk 1]
```diff
@@ -335,12 +336,23 @@ def _get_cnt_info(input_fname, eog, ecg, emg, misc, data_format, date_format, he
...
                 n_bytes = 2
                 n_samples = data_size // (n_bytes * n_channels)
+                annotations = _read_annotations_cnt(input_fname, data_format="int16")
+                if annotations.onset[-1] * sfreq > n_samples:
+                    n_bytes = 4
+                    n_samples = n_samples_header
+                    warn(
+                        "Annotations are outside data range. "
+                        "Changing data format to 'int32'."
+                    )
```
Comment (withmywoessner) [inline mne/io/cnt/cnt.py:346]: @larsoner I added a check to see if there were any annotations outside the actual amount of data. If there are, I switch the data_format. I could not figure out a way to detect the correct number of samples with certainty. It only triggers when using the `'auto'` data_format which already doesn't work with the current testing data I uploaded. Is this fine with you?

File: mne/io/cnt/tests/test_cnt.py [hunk 2]
```diff
@@ -86,6 +87,16 @@ def test_compare_events_and_annotations():
...
+@testing.requires_testing_data
+def verify_reading_bytes():
+    raw_16 = read_raw_cnt(fname, preload=True)
+    raw_32 = read_raw_cnt(fname_bad_spans, preload=True)
+
+    # Verify that the number of bytes read is correct
+    assert len(raw_16) == 3070
+    assert len(raw_32) == 90000
+
+
```
Comment (withmywoessner) [inline mne/io/cnt/tests/test_cnt.py:99]: For some reason I was unable to run my test locally using pytest

**Comment by drammock** [comment]
Test function names must start with `test_` yours starts with `verify_`

-------- Original Message --------

**Review by larsoner — COMMENTED** [review]
And don't forget to change the hash and version number of the testing dataset in `mne/datasets/config.py`

**Comment by withmywoessner** [comment]
Done! @larsoner New hash and version is in the config file:
image
Everything should be good to go! Thanks!

**Comment by larsoner** [comment]
CIs are unhappy, can you look?

**Comment by withmywoessner** [comment]
> CIs are unhappy, can you look?

I forgot a docustring. It should hopefully be good @larsoner. I think this is casuing it to fail now:
https://github.com/mne-tools/mne-python/blob/1ffe426fb1ecbc36a562b0aaa9e337000bca0dd2/mne/io/cnt/tests/test_cnt.py#L23

**Comment by larsoner** [comment]
Looking at the logs the error is:
```
mne/io/cnt/tests/test_cnt.py:57: in test_auto_data
    with pytest.warns(RuntimeWarning, match="Omitted 6 annot"):
E   RuntimeWarning: Could not define the number of bytes automatically. Defaulting to 2.
```
So I think you need to change the test now to catch a different warning string (or maybe multiple warning strings with multiple `pytest.warns`)

**Comment by withmywoessner** [comment]
> Looking at the logs the error is:
> 
> ```
> mne/io/cnt/tests/test_cnt.py:57: in test_auto_data
>     with pytest.warns(RuntimeWarning, match="Omitted 6 annot"):
> E   RuntimeWarning: Could not define the number of bytes automatically. Defaulting to 2.
> ```
> 
> So I think you need to change the test now to catch a different warning string (or maybe multiple warning strings with multiple `pytest.warns`)

Oh, I thought it was the `_no_parse` test causing it to fail. I saw this when looking at the log:
image
I removed it and they pass now. @larsoner

**Comment by larsoner** [comment]
Are you sure tests are passing and not being skipped at your end? In your mne-testing-data PR you had an outdated version in your `version.txt` locally, maybe that's causing a problem. Or maybe you're not on `pytest>=8` where they changed the warning catching behavior. In any case, I could replicate the error locally and pushed a commit to fix it.

**Comment by larsoner** [comment]
@withmywoessner ready to go from your end? I clicked the "ready for review" button in case but let me know either way (and no need to put `[wip]` in your title, it's clear from draft status!)

**Comment by withmywoessner** [comment]
> Or maybe you're not on `pytest>=8` where they changed the warning catching behavior. In any case, I could replicate the error locally and pushed a commit to fix it.

This fixed it. Thank you!

**Comment by withmywoessner** [comment]
> @withmywoessner ready to go from your end? I clicked the "ready for review" button in case but let me know either way (and no need to put `[wip]` in your title, it's clear from draft status!)

I am @larsoner. Thanks!

File: doc/changes/devel/12393.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Change how samples are read using :func:`mne.io.read_raw_cnt`
```
Comment (larsoner) [inline doc/changes/devel/12393.bugfix.rst:1]:
```suggestion
Change how samples are read when using ``data_format='auto'`` in :func:`mne.io.read_raw_cnt`, by `Jacob Wossner`_.
```

**Review by larsoner — APPROVED** [review]
Just one minor tweak that I'll commit, thanks @withmywoessner !

**Pull request merged**